Database: https://www.kaggle.com/datasets/bkcoban/retail-transaction-dataset

Very simple dataset with 30,000 entries that follows the style seen in the lecture

In [ ]:
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)
#fix for an error with mlxtend package showing depreceation warnings
#run before running code below

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
df = pd.read_csv('products.csv')
df = df[['Products']]
df.head()
#reads products.csv and takes the products column to try and match what we see in the example
#must download csv from kaggle and use google colabs upload function to upload

,Products
0,"Dish Sponge, Flatbread with Meat, Chips, Orang..."
1,"Onion, Juice, Flatbread with Meat, Chicken"
2,"Egg, Flatbread with Meat, Banana, Pizza"
3,"Ice Cream, Soda, Orange, Potato, Cereal, Choco..."
4,"Ice Cream, Soap, Shampoo, Chicken, Banana, Bea..."


In [ ]:
df.shape
#30,000 entries in products column

(30000, 1)

In [ ]:
data=list(df['Products'].apply(lambda x:x.split(",")))
data
#splits strings into a list using comma as delimiter

[['Dish Sponge',
  ' Flatbread with Meat',
  ' Chips',
  ' Orange',
  ' Butter',
  ' Sausage',
  ' Cucumber',
  ' Rice',
  ' Ice Cream'],
 ['Onion', ' Juice', ' Flatbread with Meat', ' Chicken'],
 ['Egg', ' Flatbread with Meat', ' Banana', ' Pizza'],
 ['Ice Cream',
  ' Soda',
  ' Orange',
  ' Potato',
  ' Cereal',
  ' Chocolate',
  ' Cracker'],
 ['Ice Cream',
  ' Soap',
  ' Shampoo',
  ' Chicken',
  ' Banana',
  ' Beans',
  ' Cheese'],
 ['Juice',
  ' Shampoo',
  ' Bread',
  ' Flatbread with Meat',
  ' Cracker',
  ' Cucumber',
  ' Fish'],
 ['Soda',
  ' Chocolate',
  ' Beans',
  ' Juice',
  ' Orange',
  ' Bread',
  ' Tomato',
  ' Yogurt',
  ' Minced Meat'],
 ['Cookie',
  ' Cereal',
  ' Sausage',
  ' Chips',
  ' Chocolate',
  ' Lentil',
  ' Water',
  ' Egg',
  ' Cheese'],
 ['Soda', ' Rice', ' Soap', ' Chickpeas'],
 ['Orange', ' Minced Meat', ' Chickpeas'],
 ['Milk',
  ' Minced Meat',
  ' Chips',
  ' Chicken',
  ' Onion',
  ' Cheese',
  ' Chickpeas',
  ' Cereal'],
 ['Dish Sponge', ' Onion'

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)
df=df.astype(int)
df
#one hot encoding to label items that appear in each transaction as 1 and empty ones 0

,Apple,Banana,Beans,Bread,Butter,Cereal,Cheese,Chicken,Chickpeas,Chips,...,Potato,Rice,Sausage,Shampoo,Soap,Soda,Strawberry,Tomato,Water,Yogurt
0,0,0,0,0,1,0,0,0,0,1,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,1,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,1,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,1,1,0,0,0,1,1,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29995,0,0,1,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,0
29996,0,0,0,0,0,0,0,0,0,0,...,0,0,1,0,0,0,0,0,0,0
29997,0,0,0,0,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
29998,0,0,0,0,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
df=apriori(df,min_support=0.03,use_colnames=True)
#sets support value at 0.03 due to the large datset, support values are low

In [ ]:
df

,support,itemsets
0,0.136667,( Apple)
1,0.138333,( Banana)
2,0.140000,( Beans)
3,0.135833,( Bread)
4,0.136833,( Butter)
...,...,...
855,0.019500,"( Strawberry, Water)"
856,0.019767,"( Strawberry, Yogurt)"
857,0.018200,"( Tomato, Water)"
858,0.018267,"( Tomato, Yogurt)"


In [ ]:
df_ar=association_rules(df,metric='confidence', min_threshold=0.2)
#sets the confidence threshold to 0.48; only associations with confidence > .48 are taken

In [ ]:
df_ar
#shows associations with a confidence level of 0.2 or above

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,leverage,conviction,zhangs_metric
0,( Butter),( Cereal),0.136833,0.207833,0.030300,0.221437,1.065456,0.001861,1.017473,0.071174
1,( Chips),( Cereal),0.136700,0.207833,0.030000,0.219459,1.055936,0.001589,1.014894,0.061361
2,( Cookie),( Cereal),0.138367,0.207833,0.030133,0.217779,1.047853,0.001376,1.012714,0.053002
3,( Ice Cream),( Cereal),0.142033,0.207833,0.030233,0.212861,1.024190,0.000714,1.006387,0.027529
4,( Juice),( Cereal),0.142300,0.207833,0.030200,0.212228,1.021144,0.000625,1.005578,0.024141
5,( Cereal),( Milk),0.207833,0.137267,0.076933,0.370168,2.696710,0.048405,1.369784,0.794249
6,( Milk),( Cereal),0.137267,0.207833,0.076933,0.560466,2.696710,0.048405,1.802289,0.729284
7,( Onion),( Cereal),0.135800,0.207833,0.030867,0.227295,1.093641,0.002643,1.025186,0.099078
8,( Soda),( Cereal),0.140567,0.207833,0.031100,0.221247,1.064542,0.001886,1.017225,0.070545


###3. Provide a few paragraphs interpreting the results for your boss (easy reading narration)



This analysis has 30000 retail transactions, after looking at we noticed some interesting connection between products that can help us to improve our business. The strongest link we found in data is between Cereal and Milk. When customer buys cereal there is 37% chance that they will also buy milk and when customer buy milk there is 56% chance that they will buy cereal. If we look at the lift value which is 2.7 shows that customers more likely to buy these items together.
However, if we look on other data they have connection like butter with cereal, chips with cereal, cookie with cereal, Juice with cereal, etc. but dont have as strong connection as we found in between cereal and milk and also if we look at the lift value which also close to 1. From the business perspective we can focus on the items which shows strong relations like milk and cereal. We can put them together or make some special offer like if you buy cereal and milk together is some percentage discount or online if someone buy milk suggest cereal.